BSP/STM32F413/Src/BSP_Contactor.c

```diff
@@ -4,61 +4,131 @@
+	// PA5 - input
+	Setup(C2_PORT, &GPIO_C2Init, 
+			GPIO_Pin_4, GPIO_Pin_5, GPIO_Speed_50MHz,
+			GPIO_OType_PP, GPIO_PuPd_UP);
+
+	// Initialize Port C for Contactor 3
+	// PC0 - output
+	// PC1 - input
+	Setup(C3_PORT, &GPIO_C3Init, 
+			GPIO_Pin_0, GPIO_Pin_1, GPIO_Speed_50MHz,
+			GPIO_OType_PP, GPIO_PuPd_UP);
 }
 
 /**
- * @brief   Closes the Contactor switch i.e. turns on the whole electrical system.
+ * @brief   Closes the specified Contactor switch(es) i.e. turns on the whole electrical system.
  * @note    May be good in the future to make this return something if the contactor could not successfully close.
- * @param   None
+ * @param   Contactor to turn on
  * @return  None
  */
-void BSP_Contactor_On(void) {
-    GPIO_WriteBit(GPIOB, GPIO_Pin_0, Bit_SET);      // Set pin 0 high
+void BSP_Contactor_On(CONT_CHOICE contactorChoice) {
+	// set output pins HIGH
+	if (contactorChoice == ARRAY_CONTACTOR) {
+		GPIO_WriteBit(C1_PORT, GPIO_Pin_0, Bit_SET);
+	}
+	if (contactorChoice == HVHIGH_CONTACTOR) {
+		GPIO_WriteBit(C2_PORT, GPIO_Pin_4, Bit_SET);
+	}
+	if (contactorChoice == HVLOW_CONTACTOR) {
+		GPIO_WriteBit(C3_PORT, GPIO_Pin_0, Bit_SET);
+	}
+	if (contactorChoice == ALL_CONTACTORS) {
+		GPIO_WriteBit(C1_PORT, GPIO_Pin_0, Bit_SET);
+		GPIO_WriteBit(C2_PORT, GPIO_Pin_4, Bit_SET);
+		GPIO_WriteBit(C3_PORT, GPIO_Pin_0, Bit_SET);
+	}
 }
 
 /**
- * @brief   Opens the Contactor switch i.e. turns off the whole electrical system.
- * @note    May be good in the future to make this return something if the contactor could not successfully open.
- * @param   None
+ * @brief   Opens the specified Contactor switch(es) i.e. turns off the whole electrical system.
+ * @param   Contactor to turn off
  * @return  None
  */
-void BSP_Contactor_Off(void) {
-    GPIO_WriteBit(GPIOB, GPIO_Pin_0, Bit_RESET);	// Set pin 0 low
+void BSP_Contactor_Off(CONT_CHOICE contactorChoice) {
+    // set output pins LOW
+	if (contactorChoice == ARRAY_CONTACTOR) {
+		GPIO_WriteBit(C1_PORT, GPIO_Pin_0, Bit_RESET);
+	}
+	if (contactorChoice == HVHIGH_CONTACTOR) {
+		GPIO_WriteBit(C2_PORT, GPIO_Pin_4, Bit_RESET);
+	}
+	if (contactorChoice == HVLOW_CONTACTOR) {
+		GPIO_WriteBit(C3_PORT, GPIO_Pin_0, Bit_RESET);
+	}
+	if (contactorChoice == ALL_CONTACTORS) {
+		GPIO_WriteBit(C1_PORT, GPIO_Pin_0, Bit_RESET);
+		GPIO_WriteBit(C2_PORT, GPIO_Pin_4, Bit_RESET);
+		GPIO_WriteBit(C3_PORT, GPIO_Pin_0, Bit_RESET);
+	}
 }
 
 /**
  * @brief   Gets the state of the Contactor switch from one of its AUX pins.
- * @param   None
+ * @note	You cannot get the state of ALL_CONTACTORS. As such, if that param is passed, it will return the state of the array contactor.
+ * @param   Contactor to get state of
  * @return  0 if contactor is off/open, 1 if on/closed
  */
-bool BSP_Contactor_GetState(void) {
-    // The AUX pin we have connected to is negative logic
-    if((GPIOB->IDR & GPIO_Pin_1) >> 1) {
-		return 0;
-	} else {
-		return 1;
+bool BSP_Contactor_GetState(CONT_CHOICE contactorChoice) {
```

Comment:
I think best use case for ALL_CONTACTORS should return if any of the contactors are off. That way, if anyone hits the kill switch, no matter what the kill switch is in series with, the BPS will shut down all the contactors.